### PR TITLE
Miopen dialect opt step8 : Lowering logic for `blockwise_load`/`blockwise_store`/`threadwise_load`/`threadwise_store`.

### PR DESCRIPTION
**Pull request by whchung**
Lower `blockwise_load` to `threadwise_load`, `blockwise_store` to `threadwise_store`.

### CODE REVIEW

**Comment by whchung** [comment]
After the deep dive with @asroy today, some observations:

- need to break all vector types to scalar types. vectors would only be used transiently in case we wish to actually emit vector load / store instructions. tuples would only carry scalars.
- load/store take place on top-level tensor coordinates, depending on the layout, a vector load might trigger scalar stores, and scalar loads might trigger vector stores.

**Comment by whchung** [comment]
jenkins: retest this please.

**Comment by whchung** [comment]
Adding @asleepzzz as I have to change OOB store logic to switch from `i32` to `index` type to attain symmetry in the implementation.  I expect symmetry inside `emitLoadLogic` and `emitStoreLogic`.

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h [hunk 1]
```diff
@@ -363,6 +366,7 @@ class PopulateParamsBase {
       dataPerThreadCopyGemmPos2 = dataPerThreadCopyGemmVectorized;
       derived.srcVectorReadDim = 2;
     }
+    assert(derived.srcVectorReadDim != 0);
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h]: Does this assertion serve only to make sure you've set `derived.srcVectorReadDim` in each branch of the `if`? If so, would a comment be called for to make sure people know why this is here?

Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h]: The value 0,1,2 is pretty bad. I’ll fix them with constant literals

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 2]
```diff
@@ -256,23 +409,49 @@ inline void emitStoreLogic(OpBuilder &b, Location loc, MemRefType destType,
...
+      // Issue vector store.
+      if (destType.getMemorySpace() == 0) {
+        // use raw buffer store if the dest memref is on address space 0
+        SmallVector<Value, 4> destLowerIndicesI32;
+        for (auto v : destLowerIndices)
+          destLowerIndicesI32.push_back(
+              b.create<IndexCastOp>(loc, v, b.getIntegerType(32)));
+        b.create<gpu::RawbufStoreOp>(loc, value, dest, oob,
+                                     destLowerIndicesI32);
+      } else {
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:428]: consider user vector write so we could possibly emit `ds_write_b128`.

Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:428]: defer this to the next PR.

**Comment by whchung** [comment]
Current status of the PR:

- ensure all cases which doesn't require OOB checks work fine.
- ensure all cases with OOB checks, without padding dimensions along GemmK/M/N direction work fine.
- ensure all cases with OOB checks, with padding along GemmK dimension work fine.

we should be able to cover all resnet50 configs given the current condition of PR #200 . 

the final 1 case need more work:
- ensure all cases with OOB checks, with padding along GemmM/N direction work fine.

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h [hunk 3]
```diff
@@ -316,15 +320,15 @@ class PopulateParamsBase {
                                                   bool isGemmA,
                                                   DerivedParams &derived) {
 
-    bool gemmPos1Vectorizable = false;
+    bool gemmKVectorizable = false;
     int64_t vectorizableLength = 0;
     if (isGemmA) {
       obtainGemmADimKVectorizable(ctx.opType, ctx.dimIndexVal,
```
Comment (asroy) [inline mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:326]: I didn't review what's happening inside ```obtainGemmADimKVectorizable```, etc

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h [hunk 4]
```diff
@@ -341,8 +345,18 @@ class PopulateParamsBase {
...
+    // FIXME: vectorizationSize is fixed to begin from 4 for now.
+    // Change ConvolutionContext so it carries data type information
+    // so we can possibly use 8 in case of f16.
+    int64_t vectorizationSize = 4;
+    // FIXME: set vectorizationSize be 1 for backward data for now.
+    // The logic for deciding vectorization size and dimension for
+    // backward data has to be reviewed.
+    auto opType = ctx.opType;
+    if (opType == mlir::miopen::ConvOpType::Conv2DBwdDataOpType) {
+      vectorizationSize = 1;
```
Comment (asroy) [inline mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:359]: vectorizationSize is 1, so it should be safe.

This function (PopulateParamsBase) doesn't have much to be reviewed for now.

obtainGemmADimKVectorizable, etc and vectorizationSize need to be updated in future PR

**Review by asroy — APPROVED** [review]

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h (hunk 4 above)
Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:359]: we will need to review those logic and file subsequent PRs.

File: mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h (hunk 3 above)
Comment (whchung) [reply mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:326]: it's okay for now. but we will need to review them in subsequent PRs.

**Comment by whchung** [comment]
jenkins: retest this please.

**Comment by whchung** [comment]
Merging the PR as it has been verified on the nightly CI:

http://ml-ci.amd.com:21096/job/MLIR/job/test-step8/5/console

**Comment by whchung** [comment]
`rocm-head-ci` has actually passed, see:
http://10.236.106.97:8080/job/mlir/job/PR-200/5/

**Pull request merged**